src/Features/CSharp/Portable/ChangeSignature/CSharpChangeSignatureService.cs

```diff
@@ -469,14 +477,24 @@ private bool InSymbolHeader(SyntaxNode matchingNode, int position)
 
             Debug.Assert(false, "Unknown reference location");
             return null;
+
+            async Task<(SemanticModel, ImmutableArray<ISymbol>)> GetSemanticModelAndRecommendations()
+            {
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var recommendations = await Recommender.GetImmutableRecommendedSymbolsAtPositionAsync(
+                    semanticModel, originalNode.SpanStart, document.Project.Solution.Workspace, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                return (semanticModel, recommendations);
+            }
         }
 
-        private T UpdateArgumentList<T>(
+        private async Task<T> UpdateArgumentListAsync<T>(
             ISymbol declarationSymbol,
             SignatureChange signaturePermutation,
             T argumentList,
             bool isReducedExtensionMethod,
-            bool isParamsArrayExpanded) where T : BaseArgumentListSyntax
+            bool isParamsArrayExpanded,
+            Func<Task<(SemanticModel, ImmutableArray<ISymbol>)>> getSemanticModelAndRecommendations) where T : BaseArgumentListSyntax
```

Comment:
this is super wonky. all your GetSemanticModelAndRecommendations function does is:

```c#
await Recommender.GetImmutableRecommendedSymbolsAtPositionAsync(
                    semanticModel, originalNode.SpanStart, document.Project.Solution.Workspace, cancellationToken: cancellationToken)
```

so it basically depends on data you could easily pass in, instead of the indirection.

Reply:
Removed indirection.